lib/encryptionmanager.cpp

```diff
@@ -0,0 +1,189 @@
+
+class EncryptionManager::Private
+{
+    public:
+        explicit Private(const QByteArray& encryptionAccountPickle, float signedKeysProportion, float oneTimeKeyThreshold)
+            : olmAccount(new Account(encryptionAccountPickle)), // TODO: passphrase even with qtkeychain?
+              signedKeysProportion(move(signedKeysProportion)),
+              oneTimeKeyThreshold(move(oneTimeKeyThreshold)),
+              targetKeysNumber(olmAccount->maxOneTimeKeys()) // 2 // see note below
+        {
+            Q_ASSERT((0 <= signedKeysProportion) && (signedKeysProportion <= 1));
+            Q_ASSERT((0 <= oneTimeKeyThreshold) && (oneTimeKeyThreshold <= 1));
+            /*
+             * Note about targetKeysNumber:
+             *
+             * From: https://github.com/Zil0/matrix-python-sdk/
+             * File: matrix_client/crypto/olm_device.py
+             *
+             * Try to maintain half the number of one-time keys libolm can hold uploaded
+             * on the HS. This is because some keys will be claimed by peers but not
+             * used instantly, and we want them to stay in libolm, until the limit is reached
+             * and it starts discarding keys, starting by the oldest.
+             */
+        }
+        ~Private()
+        {
+            delete olmAccount;
+        }
+
+        UploadKeysJob* uploadIdentityKeysJob = nullptr;
+        UploadKeysJob* uploadOneTimeKeysJob = nullptr;
+
+        Account* olmAccount;
+        const QByteArray encryptionAccountPickle;
+
+        float signedKeysProportion;
+        float oneTimeKeyThreshold;
+        int targetKeysNumber;
+
+        void updateKeysToUpload();
+        bool oneTimeKeyShouldUpload();
+
+        QHash<QString, int> oneTimeKeyCounts;
+        void setOneTimeKeyCounts(const QHash<QString, int> oneTimeKeyCountsNewValue)
+        {
+            oneTimeKeyCounts = oneTimeKeyCountsNewValue;
+            updateKeysToUpload();
+        }
+        QHash<QString, int> oneTimeKeysToUploadCounts;
+        QHash<QString, int> targetOneTimeKeyCounts
+        {
+            {signedCurve25519Name, qRound(signedKeysProportion * targetKeysNumber)},
+            {curve25519Name, qRound((1-signedKeysProportion) * targetKeysNumber)}
+        };
+};
+
+EncryptionManager::EncryptionManager(const QByteArray &encryptionAccountPickle, float signedKeysProportion, float oneTimeKeyThreshold,
+                                     QObject* parent)
+    : QObject(parent),
+    d(std::make_unique<Private>(std::move(encryptionAccountPickle), std::move(signedKeysProportion), std::move(oneTimeKeyThreshold)))
+{
+
+}
+
+EncryptionManager::~EncryptionManager()
+{
+
+}
+
+void EncryptionManager::uploadIdentityKeys(Connection* connection)
+{
+    DeviceKeys deviceKeys;
+
+    deviceKeys.userId = connection->userId();
+    deviceKeys.deviceId = connection->deviceId();
+
+    QStringList algorithms = { olmCurve25519AesSha256AlgoName,
```

Comment:
```suggestion
    static const QStringList algorithms = { olmCurve25519AesSha256AlgoName,
```
And probably just move it up to static constants with the name like `SupportedAlgorithms`.